### ISSUE TITLE
 Heartbeat failed for group xxxWorker because it is rebalancing

### ISSUE THREAD

**Issue by ealfatt** [dpkp/kafka-python#1418]
Hi all
I'm facing this problem that is driving to me crazy with 1.4.1 version of Kafka python
the instruction that i perform are:
- create a consumer in this way
  KafkaConsumer(bootstrap_servers=kafka_multi_hosts,
                                                    auto_offset_reset=earliest,
                                                    enable_auto_commit=False,
                                                    group_id=group_name,
                                                    reconnect_backoff_ms=1,
                                                    consumer_timeout_ms=5000)
- subscribe on a topics
- finally the consumer.poll(500)

no problem till now but then in the log i see:

[INFO] 03/08/2018 02:52:53 PM Subscribe executed.
[INFO] 03/08/2018 02:52:53 PM Initialization pool executed.
[INFO] 03/08/2018 02:52:53 PM Subscribed to topic: event
[INFO] 03/08/2018 02:52:53 PM eventHandle connected
[WARNING] 03/08/2018 02:53:23 PM Heartbeat failed for group emsWorker because it is rebalancing
[WARNING] 03/08/2018 02:53:26 PM Heartbeat failed for group emsWorker because it is rebalancing
[WARNING] 03/08/2018 02:53:29 PM Heartbeat failed for group emsWorker because it is rebalancing
[WARNING] 03/08/2018 02:53:32 PM Heartbeat failed for group emsWorker because it is rebalancing
......
......
......
[WARNING] 03/08/2018 02:57:48 PM Heartbeat failed for group emsWorker because it is rebalancing
[WARNING] 03/08/2018 02:57:51 PM Heartbeat failed for group emsWorker because it is rebalancing
[INFO] 03/08/2018 02:57:53 PM Leaving consumer group (group_name).

why?

I've also added the option max_poll_records=50 in the kafkaConsumer definition but nothing is changed

@dpkp can you help me?
do you know if the 1.4.1 version presents some problem about that? Because in the previous version I can not see this problem.

thanks in advance

**Comment by ealfatt** [comment]
Hi
adding an information,
I've also added the option  max_poll_records=50 in the kafkaConsumer definition but nothing is changed

**Comment by ealfatt** [comment]
Hi 
i forgot to say you that i'm using the 1.4.1 version of Kafka python

thanks

**Comment by infecto** [comment]
@ealfatt Yeah this is definitely newer behavior since there was not a background heartbeat thread so it basically happened during polling. 

I am guessing after the last message your worker is not rejoining?

**Comment by dpkp** [comment]
kafka-python 1.4 changes heartbeats to a background thread, and it enables additional timeout configurations in max_poll_interval_ms and session_timeout_ms . It is certainly possible there is a bug in the code, but all of our tests pass so I'm not aware of anything obvious. Can you enable debug logs and see if anything looks strange?

**Comment by ealfatt** [comment]
@infecto yes, after the last message my worker was not rejoining (thanks for the response)
Hi @dpkp, thanks for the response. I'll try to play with those two parameter and try to enable the debug logs (how i can do that? is it explained in the documentation?)

**Comment by ealfatt** [comment]
Hi @dpkp 
below 2 logs of what i can see when the issue above happen.

[log1.txt](https://github.com/dpkp/kafka-python/files/1811218/log1.txt)
[log2.txt](https://github.com/dpkp/kafka-python/files/1811225/log2.txt)

could yo give a look and let me know your feedback?
thanks a lot in advance

**Comment by dpkp** [comment]
This log entry: `Heartbeat poll expired, leaving group` means that your consumer application is not processing messages fast enough. What is your `max_poll_interval_ms` configuration? Is there something that might block your poll loop for longer than that amount?

**Comment by slash31** [comment]
After upgrading to 1.4.3 from 1.3.5 a couple of weeks ago, I started seeing my consumer groups rebalancing at the session timeout setting, no matter what I set it to. I tried:
- adjusting the poll/heartbeat/session timers, as well as max recs polled
- setting the assignment algorithm deterministically to round robin
- turning off auto-commit and manually committing
- upgrading Kafka to the latest available parcel

I see heartbeat requests and successful responses, but commit responses with an Error 25 just before the error's thrown that the consumer can't commit successfully:

> 2018-10-28 20:35:06,563 kafka.conn   DEBUG    <BrokerConnection node_id=772 host=xvda-c14.xv.dc.openx.org:9092 <connected> [IPv4 ('10.11.3.14', 9092)]> Response 103 (0.808000564575 ms): OffsetCommitResponse_v2(topics=[(topic=u'netflow.raw', partitions=[(partition=0, error_code=25), (partition=5, error_code=25), (partition=10, error_code=25)])])
> 2018-10-28 20:35:06,563 kafka.conn   DEBUG    <BrokerConnection node_id=772 host=xvda-c14.xv.dc.openx.org:9092 <connected> [IPv4 ('10.11.3.14', 9092)]> Response 103 (0.808000564575 ms): OffsetCommitResponse_v2(topics=[(topic=u'netflow.raw', partitions=[(partition=0, error_code=25), (partition=5, error_code=25), (partition=10, error_code=25)])])
> 2018-10-28 20:35:06,563 kafka.coordinator.consumer DEBUG    OffsetCommit for group netflow_raw failed: [Error 25] UnknownMemberIdError: netflow_raw
> 2018-10-28 20:35:06,563 kafka.coordinator.consumer DEBUG    OffsetCommit for group netflow_raw failed: [Error 25] UnknownMemberIdError: netflow_raw

I downgraded the version to 1.3.5 again and everything is happy, all consumers have been running smoothly for hours now with no issue.One other thing I noticed is a lot of wow/flutter on graphs of this data (netflow) when using 1.4.3, which smoothed out again after downgrading. The setup for each of my clusters is 10 consumers running on one host via multiprocess, consuming single topic with 10 partitions on three brokers. The application is very simple/lightweight; it reads 2k-4k messages per second depending on time of day, formats/augments some of the data, and then dumps it to CrateDB. With the 1.3.5 client, I can easy get > 20MBps throughput in this application, with each db write of 5k rows taking < 1s.

If you're interested in any of the other logs/graphs from either kafka-python or my brokers, I'd be happy to share.

**Comment by julienaubert** [comment]
I see this as well.

The problem for me: I have a set of consumers, each polling 500 records at a time, being part of group generation N

Another consumer joins. Causing a rebalance. Generation N+1

I can see in the logs that a consumer that is still processing the 500 records that it got at generation N, will log that the heart beat failed due to rebalancing. But the processing code does not learn this, so it continues processing..

I can see that this consumer, the broker doesn't include it in the rebalancing. I wonder why, as it was sending heart-beats? Is this considered a bug? 

I only quickly glimpsed at the code. It seems this could happen: HeartBeat ends up in a state "do not send more heartbeats" due to rebalancing and the broker takikng sufficiently long time to perform the rebalancing, during which time the consumer has not sent any heartbeats, so is subsequently not included. It seems some special-case handling would be needed (eg reset timeouts) when rebalancing occurs.

Finally when it commits and polls again it learns that it needs to join.. and then we get to generation N+2, and so we get this cyclic behaviour with consumers causing rebalances because they were not included.

**Comment by 88manpreet** [comment]
We have also started seeing this error quite frequently with the upgrade to 1.4.3. So far none of tuning has turned out to be helpful.

**Comment by dpkp** [comment]
Can you try 1.4.4 ? There are some fixes included that may help.

**Comment by rafenden** [comment]
It happens to me as well, I'm running `1.4.4`

**Comment by goern** [comment]
I can see it on `1.4.4` too.

**Comment by dpkp** [comment]
One point to keep in mind: heartbeats "failing" during rebalance is completely normal. The main java client treats this as a debug log, not a warning. If this is simply a case of spammy logs, we should probably return the log entry to debug.

However, if you are seeing unexpected rebalances and/or rebalances are not resolving correctly, then there may be something else going on. It is hard to say what that is from the issue detail so far. Some possibilities: your consumer is taking too long to process each batch, causing a max poll interval timeout. Or, there could be some bug or deadlock in kafka-python that is causing the consumer to stop processing records altogether. Do you have any more information on specifically what you are seeing in the consumer behavior apart from the heartbeat?

**Comment by vimal3271** [comment]
I am seeing consumer rebalances even if there is no messages to consume. Start three consumers in a group and send some messages to topic and after that stop the producer. The consumer will start seeing rebalances after 5-6mins.
Sample code here:
https://stackoverflow.com/questions/54908902/kafka-consumer-rebalance-occurs-even-if-there-is-no-message-to-consume

**Comment by dpkp** [comment]
vimal: thanks for posting. I believe you may be hitting lock contention between an idle client.poll -- which can block and hold the client lock for the entire request_timeout_ms -- and the attempt by the heartbeat thread to send a new request. It seems to me that we may need to use KafkaClient.wakeup() to make sure that the polling thread drops the lock if/when we need to send a request from a different thread.

**Comment by vimal3271** [comment]
@dpkp Thanks for the reply. Is there any workaround to avoid this?

**Comment by dpkp** [comment]
This shouldn't be an issue when messages are flowing through your topics at a steady rate. If this is just a test environment, and you expect your production environment to have more steady live data, then you could just ignore the error in testing.  But if you are managing a topic w/ very low traffic -- delays of minutes between consecutive messages, for example -- you might try to reduce the `request_timeout_ms` to something closer to the heartbeat_interval_ms, which should prevent the read side from blocking for much longer than the heartbeat timeout. But note that other timeouts may also need to change (`max_poll_interval_ms` and `session_timeout_ms` perhaps). Another workaround might be to reduce `metadata_max_age_ms` to something close / equal to your heartbeat_timeout_ms. This will cause more frequent metadata requests, but should unblock the send side when there is no socket data available for reads.

**Comment by vimal3271** [comment]
@dpkp Thanks for the workaround. The rebalance is not happening anymore. I used the second approach of reducing` metadata_max_age_ms` closer to `heartbeat_time_out`

**Comment by mhorowitz0** [comment]
> vimal: thanks for posting. I believe you may be hitting lock contention between an idle client.poll -- which can block and hold the client lock for the entire request_timeout_ms -- and the attempt by the heartbeat thread to send a new request. It seems to me that we may need to use KafkaClient.wakeup() to make sure that the polling thread drops the lock if/when we need to send a request from a different thread.

@dpkp I think I am hitting this exact issue with the lock contention, and it happens quite frequently (using 1.4.4).  I turned the log level to DEBUG, and I can see that the heartbeats were every 3 seconds, then it went into a _client.poll -> _selector.select -> _epoll.poll (confirmed in pdb stack trace) and just hung for 5 minutes until it timed out, despite the fact the log end offset was way ahead of the current offset.  During that time there were no heartbeats sent.  I suspect the session timedout after 10 seconds, and the poll was left in the lurch until it reached the request_timeout_ms, after which it recognized the heartbeat timeout and started refreshing the metadata and doing a rejoin of the consumer group.  After that the heartbeats started again every 3 seconds.  I saw another issue with the heartbeat not being started after JoinGroup, but from what I can see the heartbeats were regular since the last JoinGroup in the logs, and only stopped when it went into that poll.

The issue happens randomly, but it happens often.  The result is that consumers periodically hang in polls and messages build up and the lag becomes large, then it has to rejoin the group and then catch up on a backlog of messages.  This problem becomes even worse when the poll goes for more than the max_poll_interval_ms causing runaway constant rebalancing, especially in groups with many consumers.  I've mitigated the runaway rebalancing by just upping the max_poll_interval_ms to 1 hour, but I still see those constant poll hangs on partitions with plenty of lag, from anywhere from 20 seconds to 5 minutes.  I can only thus far attribute them to the stoppage of the heartbeats and a session timeout as a result.

I was thinking about reducing the request_timeout_ms from the default of 5 minutes to 30 seconds, but that will only work around the issue, it seems the underlying root cause of those heartbeats stopping would still be there.

**Comment by jeffwidman** [comment]
@mhorowitz0 / @vimal3271 there were a number of fixes merged to `master` in the past few days that should alleviate these lock problems w/o requiring custom configs... do you mind trying them and see whether they fix these issues?

**Comment by jeffwidman** [comment]
I'm going to close, please try the new 1.4.6 release, if you still see issues please open a new ticket.

**Comment by mhorowitz0** [comment]
@jeffwidman I tried 1.4.5, and I still see the same symptoms.  I will try 1.4.6.

**Comment by mhorowitz0** [comment]
@jeffwidman I am seeing the same issue with 1.4.6.  The issue is intermittent, so it is hard to catch, but from what I can see even with a single consumer group, a poll() can hang inexplicably for several minutes even when the topic/partition has a large lag, and even in as little as 12 seconds (enough to trigger a session timeout), and during that time heartbeats stop being sent for longer than the session timeout, and so a rebalance is triggered.  I don't know why it hangs so long in the poll() when most of the time polls() take less than a second, and I don't know why the heartbeats stop, but I can only guess it is the same root cause, and most likely the lock contention @dpkp refers to above.

**Comment by artemistomaras** [comment]
I am having the same issue as @mhorowitz0 with **1.4.6** with the default provided settings for the consumer. 

A consumer inside a ConsumerGroup fails with the following log when there are no messages in the topic he is currently assigned to. 

```
2019 Apr 10 06:22:03  | 2019-04-10 06:22:02,491 - kafka.coordinator - INFO - (Re-)joining group NSFW_CONSUMER
2019 Apr 10 06:22:06  | 2019-04-10 06:22:05,582 - kafka.coordinator - INFO - Successfully joined group NSFW_CONSUMER with generation 220
2019 Apr 10 06:22:06  | 2019-04-10 06:22:05,583 - kafka.consumer.subscription_state - INFO - Updated partition assignment: [TopicPartition(topic='request_for_analyze', partition=3)]
2019 Apr 10 06:22:06  | 2019-04-10 06:22:05,583 - kafka.coordinator.consumer - INFO - Setting newly assigned partitions {TopicPartition(topic='request_for_analyze', partition=3)} for group NSFW_CONSUMER
2019 Apr 10 06:27:12  | 2019-04-10 06:25:37,880 - kafka.coordinator - WARNING - Heartbeat session expired, marking coordinator dead
2019 Apr 10 06:27:12  | 2019-04-10 06:25:37,880 - kafka.coordinator - WARNING - Marking the coordinator dead (node 0) for group NSFW_CONSUMER: Heartbeat session expired.
2019 Apr 10 06:27:12  | 2019-04-10 06:25:37,889 - kafka.cluster - INFO - Group coordinator for NSFW_CONSUMER is BrokerMetadata(nodeId=0, host='tot-kafka-0.tot-kafka-headless.tot', port=9092, rack=None)
2019 Apr 10 06:27:12  | 2019-04-10 06:25:42,893 - kafka.coordinator - WARNING - Marking the coordinator dead (node 0) for group NSFW_CONSUMER: [Error 7] RequestTimedOutError.
2019 Apr 10 06:27:12  | 2019-04-10 06:25:37,890 - kafka.coordinator - INFO - Discovered coordinator 0 for group NSFW_CONSUMER
2019 Apr 10 06:27:12  | 2019-04-10 06:25:42,897 - kafka.cluster - INFO - Group coordinator for NSFW_CONSUMER is BrokerMetadata(nodeId=0, host='tot-kafka-0.tot-kafka-headless.tot', port=9092, rack=None)
2019 Apr 10 06:27:12  | 2019-04-10 06:25:42,894 - kafka.coordinator - WARNING - Heartbeat failed for group NSFW_CONSUMER because it is rebalancing
2019 Apr 10 06:27:12  | 2019-04-10 06:25:42,893 - kafka.coordinator.consumer - WARNING - Auto offset commit failed for group NSFW_CONSUMER: [Error 7] RequestTimedOutError: NSFW_CONSUMER
2019 Apr 10 06:27:12  | 2019-04-10 06:25:42,898 - kafka.coordinator - INFO - Discovered coordinator 0 for group NSFW_CONSUMER
2019 Apr 10 06:27:12  | 2019-04-10 06:25:48,003 - kafka.cluster - INFO - Group coordinator for NSFW_CONSUMER is BrokerMetadata(nodeId=0, host='tot-kafka-0.tot-kafka-headless.tot', port=9092, rack=None)
2019 Apr 10 06:27:12  | 2019-04-10 06:25:47,901 - kafka.coordinator - WARNING - Marking the coordinator dead (node 0) for group NSFW_CONSUMER: [Error 7] RequestTimedOutError.
2019 Apr 10 06:27:12  | 2019-04-10 06:25:53,149 - kafka.coordinator - WARNING - Marking the coordinator dead (node 0) for group NSFW_CONSUMER: [Error 7] RequestTimedOutError.
2019 Apr 10 06:27:12  | 2019-04-10 06:25:48,003 - kafka.coordinator - INFO - Discovered coordinator 0 for group NSFW_CONSUMER
2019 Apr 10 06:27:12  | 2019-04-10 06:25:53,252 - kafka.cluster - INFO - Group coordinator for NSFW_CONSUMER is BrokerMetadata(nodeId=0, host='tot-kafka-0.tot-kafka-headless.tot', port=9092, rack=None)
2019 Apr 10 06:27:12  | 2019-04-10 06:25:58,254 - kafka.coordinator - WARNING - Marking the coordinator dead (node 0) for group NSFW_CONSUMER: [Error 7] RequestTimedOutError.
2019 Apr 10 06:27:12  | 2019-04-10 06:25:53,252 - kafka.coordinator - INFO - Discovered coordinator 0 for group NSFW_CONSUMER
2019 Apr 10 06:27:12  | 2019-04-10 06:25:58,358 - kafka.coordinator - INFO - Discovered coordinator 0 for group NSFW_CONSUMER
2019 Apr 10 06:27:12  | 2019-04-10 06:25:58,358 - kafka.cluster - INFO - Group coordinator for NSFW_CONSUMER is BrokerMetadata(nodeId=0, host='tot-kafka-0.tot-kafka-headless.tot', port=9092, rack=None)
2019 Apr 10 06:27:12  | 2019-04-10 06:26:00,382 - kafka.coordinator - INFO - (Re-)joining group NSFW_CONSUMER
2019 Apr 10 06:27:12  | 2019-04-10 06:26:00,485 - kafka.coordinator - INFO - Successfully joined group NSFW_CONSUMER with generation 221
2019 Apr 10 06:27:12  | 2019-04-10 06:26:00,381 - kafka.coordinator.consumer - INFO - Revoking previously assigned partitions {TopicPartition(topic='request_for_analyze', partition=3)} for group NSFW_CONSUMER
2019 Apr 10 06:27:12  | 2019-04-10 06:26:00,486 - kafka.coordinator.consumer - INFO - Setting newly assigned partitions {TopicPartition(topic='request_for_analyze', partition=3)} for group NSFW_CONSUMER
2019 Apr 10 06:27:12  | 2019-04-10 06:26:00,486 - kafka.consumer.subscription_state - INFO - Updated partition assignment: [TopicPartition(topic='request_for_analyze', partition=3)]
2019 Apr 10 06:34:42  | 2019-04-10 06:34:42,897 - kafka.client - INFO - Closing idle connection 2, last active 540000 ms ago
2019 Apr 10 06:34:42  | 2019-04-10 06:34:42,898 - kafka.conn - INFO - <BrokerConnection node_id=2 host=tot-kafka-2.tot-kafka-headless.tot:9092 <connected> [IPv4 ('10.244.3.12', 9092)]>: Closing connection. 
2019 Apr 10 07:21:58  | 2019-04-10 07:21:57,707 - kafka.coordinator - WARNING - Heartbeat failed for group NSFW_CONSUMER because it is rebalancing
2019 Apr 10 07:21:58  | 2019-04-10 07:21:57,717 - kafka.coordinator.consumer - INFO - Revoking previously assigned partitions {TopicPartition(topic='request_for_analyze', partition=3)} for group NSFW_CONSUMER
2019 Apr 10 07:21:58  | 2019-04-10 07:21:57,718 - kafka.coordinator - INFO - (Re-)joining group NSFW_CONSUMER
2019 Apr 10 07:21:58  | 2019-04-10 07:21:57,990 - kafka.coordinator - INFO - Successfully joined group NSFW_CONSUMER with generation 222
2019 Apr 10 07:21:58  | 2019-04-10 07:21:57,994 - kafka.consumer.subscription_state - INFO - Updated partition assignment: [TopicPartition(topic='request_for_analyze', partition=3)]
2019 Apr 10 07:21:58  | 2019-04-10 07:21:57,996 - kafka.coordinator.consumer - INFO - Setting newly assigned partitions {TopicPartition(topic='request_for_analyze', partition=3)} for group NSFW_CONSUMER
2019 Apr 10 07:26:02  | 2019-04-10 07:26:02,457 - kafka.coordinator - WARNING - Heartbeat failed for group NSFW_CONSUMER because it is rebalancing
2019 Apr 10 07:26:02  | 2019-04-10 07:26:02,483 - kafka.coordinator - INFO - (Re-)joining group NSFW_CONSUMER
2019 Apr 10 07:26:02  | 2019-04-10 07:26:02,482 - kafka.coordinator.consumer - INFO - Revoking previously assigned partitions {TopicPartition(topic='request_for_analyze', partition=3)} for group NSFW_CONSUMER
2019 Apr 10 07:26:02  | 2019-04-10 07:26:02,802 - kafka.coordinator - INFO - Successfully joined group NSFW_CONSUMER with generation 223
2019 Apr 10 07:26:02  | 2019-04-10 07:26:02,802 - kafka.consumer.subscription_state - INFO - Updated partition assignment: [TopicPartition(topic='request_for_analyze', partition=2)]
2019 Apr 10 07:26:02  | 2019-04-10 07:26:02,803 - kafka.coordinator.consumer - INFO - Setting newly assigned partitions {TopicPartition(topic='request_for_analyze', partition=2)} for group NSFW_CONSUMER
2019 Apr 10 07:36:02  | 2019-04-10 07:36:02,635 - kafka.coordinator - WARNING - Heartbeat session expired, marking coordinator dead
2019 Apr 10 07:36:02  | 2019-04-10 07:36:02,635 - kafka.coordinator - WARNING - Marking the coordinator dead (node 0) for group NSFW_CONSUMER: Heartbeat session expired.
2019 Apr 10 07:36:02  | 2019-04-10 07:36:02,643 - kafka.cluster - INFO - Group coordinator for NSFW_CONSUMER is BrokerMetadata(nodeId=0, host='tot-kafka-0.tot-kafka-headless.tot', port=9092, rack=None)
2019 Apr 10 07:36:02  |             rebalanced and assigned the partitions to another member.
2019 Apr 10 07:36:02  | 2019-04-10 07:36:02,644 - kafka.coordinator - INFO - Discovered coordinator 0 for group NSFW_CONSUMER
2019 Apr 10 07:36:02  | 2019-04-10 07:36:02,648 - kafka.coordinator.consumer - WARNING - Auto offset commit failed for group NSFW_CONSUMER: CommitFailedError: Commit cannot be completed since the group has already
2019 Apr 10 07:36:02  |             This means that the time between subsequent calls to poll()
2019 Apr 10 07:36:02  |             typically implies that the poll loop is spending too much
2019 Apr 10 07:36:02  |             was longer than the configured max_poll_interval_ms, which
2019 Apr 10 07:36:02  |             increasing the rebalance timeout with max_poll_interval_ms,
2019 Apr 10 07:36:02  |             time message processing. You can address this either by
2019 Apr 10 07:36:02  |             with max_poll_records.
2019 Apr 10 07:36:02  |             or by reducing the maximum size of batches returned in poll()
2019 Apr 10 07:36:02  |             
2019 Apr 10 07:36:02  | 2019-04-10 07:36:02,649 - kafka.coordinator.consumer - ERROR - Offset commit failed: This is likely to cause duplicate message delivery
2019 Apr 10 07:36:02  | Traceback (most recent call last):
  File "/opt/app/.venv/lib/python3.6/site-packages/kafka/coordinator/consumer.py", line 521, in _maybe_auto_commit_offsets_sync
    self.commit_offsets_sync(self._subscription.all_consumed_offsets())
  File "/opt/app/.venv/lib/python3.6/site-packages/kafka/coordinator/consumer.py", line 514, in commit_offsets_sync
    raise future.exception # pylint: disable-msg=raising-bad-type
kafka.errors.CommitFailedError: CommitFailedError: Commit cannot be completed since the group has already
2019 Apr 10 07:36:02  |             rebalanced and assigned the partitions to another member.
2019 Apr 10 07:36:02  |             This means that the time between subsequent calls to poll()
2019 Apr 10 07:36:02  |             was longer than the configured max_poll_interval_ms, which
2019 Apr 10 07:36:02  |             typically implies that the poll loop is spending too much
2019 Apr 10 07:36:02  |             time message processing. You can address this either by
2019 Apr 10 07:36:02  |             or by reducing the maximum size of batches returned in poll()
2019 Apr 10 07:36:02  |             increasing the rebalance timeout with max_poll_interval_ms,
2019 Apr 10 07:36:02  |             
2019 Apr 10 07:36:02  |             with max_poll_records.
2019 Apr 10 07:36:02  | 2019-04-10 07:36:02,649 - kafka.coordinator.consumer - INFO - Revoking previously assigned partitions {TopicPartition(topic='request_for_analyze', partition=2)} for group NSFW_CONSUMER
2019 Apr 10 07:36:02  | 2019-04-10 07:36:02,650 - kafka.coordinator - INFO - (Re-)joining group NSFW_CONSUMER
2019 Apr 10 07:36:03  | 2019-04-10 07:36:03,920 - kafka.coordinator - INFO - Successfully joined group NSFW_CONSUMER with generation 225
2019 Apr 10 07:36:03  | 2019-04-10 07:36:03,921 - kafka.coordinator.consumer - INFO - Setting newly assigned partitions {TopicPartition(topic='request_for_analyze', partition=4)} for group NSFW_CONSUMER
2019 Apr 10 07:36:03  | 2019-04-10 07:36:03,920 - kafka.consumer.subscription_state - INFO - Updated partition assignment: [TopicPartition(topic='request_for_analyze', partition=4)]
2019 Apr 10 07:36:04  | 2019-04-10 07:36:03,989 - kafka.conn - INFO - <BrokerConnection node_id=2 host=tot-kafka-2.tot-kafka-headless.tot:9092 <connecting> [IPv4 ('10.244.3.12', 9092)]>: connecting to tot-kafka-2.tot-kafka-headless.tot:9092 [('10.244.3.12', 9092) IPv4]
2019 Apr 10 07:36:04  | 2019-04-10 07:36:03,991 - kafka.conn - INFO - <BrokerConnection node_id=2 host=tot-kafka-2.tot-kafka-headless.tot:9092 <connecting> [IPv4 ('10.244.3.12', 9092)]>: Connection complete.
2019 Apr 10 07:43:54  | 2019-04-10 07:43:54,668 - kafka.coordinator - WARNING - Heartbeat failed for group NSFW_CONSUMER because it is rebalancing
2019 Apr 10 07:43:54  | 2019-04-10 07:43:54,677 - kafka.coordinator.consumer - INFO - Revoking previously assigned partitions {TopicPartition(topic='request_for_analyze', partition=4)} for group NSFW_CONSUMER
2019 Apr 10 07:43:54  | 2019-04-10 07:43:54,700 - kafka.coordinator - INFO - Successfully joined group NSFW_CONSUMER with generation 226
2019 Apr 10 07:43:54  | 2019-04-10 07:43:54,684 - kafka.coordinator - INFO - Elected group leader -- performing partition assignments using range
2019 Apr 10 07:43:54  | 2019-04-10 07:43:54,678 - kafka.coordinator - INFO - (Re-)joining group NSFW_CONSUMER
2019 Apr 10 07:43:54  | 2019-04-10 07:43:54,701 - kafka.coordinator.consumer - INFO - Setting newly assigned partitions {TopicPartition(topic='request_for_analyze', partition=4)} for group NSFW_CONSUMER
2019 Apr 10 07:43:54  | 2019-04-10 07:43:54,700 - kafka.consumer.subscription_state - INFO - Updated partition assignment: [TopicPartition(topic='request_for_analyze', partition=4)]
```

**Comment by jeffwidman** [comment]
Thanks for the update @mhorowitz0 and @artemistomaras, re-opening.

If anyone can capture `DEBUG` logs, this would also be helpful for tracking this down.

**Comment by artemistomaras** [comment]
@jeffwidman as requested, I am attaching the DEBUG logs from the starting of the Consumer until the error occurs

[logs.txt](https://github.com/dpkp/kafka-python/files/3072187/logs.txt)

**Comment by mhorowitz0** [comment]
@jeffwidman  Unfortunately, I can't send logs from where I work, and I can only post here from my personal computer sans any data.  It is hard for me to reproduce on demand and the last debug logs I captured are from 1.4.4, with logs on DEBUG across 100s of processes that would overflow before I could get to them by the time I notice it.   

I looked at @artemistomaras logs and it looks pretty similar.
Here is a scenario similar to mine:

Last successful heartbeats in the logs every 3 seconds like normal default:

```
2019-04-12 07:50:48,097 - kafka.coordinator - DEBUG - Received successful heartbeat response for group EXTRACT_LABELS
2019-04-12 07:50:51,145 - kafka.coordinator - DEBUG - Received successful heartbeat response for group EXTRACT_LABELS
```

Then you see here right when the next heartbeat is supposed to happen, it hangs at 07:50:54,195 at processing this FetchResponse_v4 for 53 seconds. 

```
2019-04-12 07:50:54,195 - kafka.protocol.parser - DEBUG - Received correlation id: 354
2019-04-12 07:50:54,195 - kafka.protocol.parser - DEBUG - Processing response FetchResponse_v4
2019-04-12 07:50:54,195 - kafka.conn - DEBUG - <BrokerConnection node_id=0 host=tot-kafka-0.tot-kafka-headless.tot:9092 <connected> [IPv4 ('10.244.2.226', 9092)]> Response 354 (509.2189311981201 ms): FetchResponse_v4(throttle_time_ms=0, topics=[(topics='request_for_analyze', partitions=[(partition=2, error_code=0, highwater_offset=116, last_stable_offset=-1, aborted_transactions=NULL, message_set=b'')])])
2019-04-12 07:51:47,941 - kafka.coordinator.consumer - DEBUG - Sending offset-commit request with {TopicPartition(topic='request_for_analyze', partition=2): OffsetAndMetadata(offset=116, metadata='')} for group EXTRACT_LABELS to 1
2019-04-12 07:51:47,942 - kafka.protocol.parser - DEBUG - Sending request OffsetCommitRequest_v2(consumer_group='EXTRACT_LABELS', consumer_group_generation_id=865, consumer_id='kafka-python-1.4.6-a49d7d71-4958-4249-804c-c741039872cd', retention_time=-1, topics=[(topic='request_for_analyze', partitions=[(partition=2, offset=116, metadata='')])])
2019-04-12 07:51:47,942 - kafka.conn - DEBUG - <BrokerConnection node_id=1 host=tot-kafka-1.tot-kafka-headless.tot:9092 <connected> [IPv4 ('10.244.1.207', 9092)]> Request 2215: OffsetCommitRequest_v2(consumer_group='EXTRACT_LABELS', consumer_group_generation_id=865, consumer_id='kafka-python-1.4.6-a49d7d71-4958-4249-804c-c741039872cd', retention_time=-1, topics=[(topic='request_for_analyze', partitions=[(partition=2, offset=116, metadata='')])])
2019-04-12 07:51:47,945 - kafka.protocol.parser - DEBUG - Received correlation id: 2215
2019-04-12 07:51:47,945 - kafka.protocol.parser - DEBUG - Processing response OffsetCommitResponse_v2
2019-04-12 07:51:47,945 - kafka.conn - DEBUG - <BrokerConnection node_id=1 host=tot-kafka-1.tot-kafka-headless.tot:9092 <connected> [IPv4 ('10.244.1.207', 9092)]> Response 2215 (2.673625946044922 ms): OffsetCommitResponse_v2(topics=[(topic='request_for_analyze', partitions=[(partition=2, error_code=25)])])
2019-04-12 07:51:47,946 - kafka.coordinator.consumer - DEBUG - OffsetCommit for group EXTRACT_LABELS failed: [Error 25] UnknownMemberIdError: EXTRACT_LABELS
2019-04-12 07:51:47,946 - kafka.coordinator.consumer - WARNING - Auto offset commit failed for group EXTRACT_LABELS: CommitFailedError: Commit cannot be completed since the group has already
            rebalanced and assigned the partitions to another member.
            This means that the time between subsequent calls to poll()
            was longer than the configured max_poll_interval_ms, which
            typically implies that the poll loop is spending too much
            time message processing. You can address this either by
            increasing the rebalance timeout with max_poll_interval_ms,
            or by reducing the maximum size of batches returned in poll()
            with max_poll_records.
```

I don't know this code is doing, but in my case, I have additional logging/pdb telling me during that pause it is stuck waiting for a return from a poll() via an iterator, i.e. I have logging before and after the loop condition on the iterator and one lucky time caught it and confirmed in pdb stack trace it was stuck here: _client.poll -> _selector.select -> _epoll.poll.  In my case, I know the lag on the partition had many messages, so there were messages waiting to be consumed.  I have gotten hanging anywhere from 19 seconds to 5 minutes that have caused this.  I also can confirm I have gotten this for a consumer group with a single consumer, so the rebalance is NOT due to a timeout of a different consumer in the group.

None the less, there is no heartbeat as scheduled at 07:50:54, and there is none for that entire 53 second interval, well beyond the default session timeout of 10 seconds.  So, it is no surprise when it comes out of the poll() finally the session has timed out and it has to re-join the group.  The fact that I generally see a max of 5 minutes for a pause makes sense with the default request timeout of around 5 mins.  Why it pauses for so long in a poll() when there are messages to read is a mystery to me.

The fact that both the heartbeat thread stops during that period and the poll() hangs, leads me to believe it is some lock contention between those two threads as suggested by @dpkp earlier, but that is just my guess based on my limited familiarity with the code.

**Comment by artemistomaras** [comment]
Any news on this issue?

**Comment by ottomata** [comment]
At Wikimedia, we recently upgraded from 1.4.3 because one of our services was experiencing #1590. After upgrading, we then began to experience this bug.  Or at least, we started experiencing it severely enough to cause serious consumer lag.  It is possible the missed heartbeats & rebalances happened before too, but they didn't cause serious lag.

We decided to try backporting the fix in #1590 on top of of 1.4.3, but continued to experience consumer lag due to rebalances. Once we reverted to the original 1.4.3 without #1590, the lag stopped.  I still see missed heartbeats & rebalances, but they must be happening with less frequency and/or resolving quickly enough to keep consumption steady.

Is it possible that #1590 contributed to this bug?

Ping @elukey

**Comment by elukey** [comment]
> At Wikimedia, we recently upgraded from 1.4.3 because one of our services was experiencing #1628. After upgrading, we then began to experience this bug. Or at least, we started experiencing it severely enough to cause serious consumer lag. It is possible the missed heartbeats & rebalances happened before too, but they didn't cause serious lag.
> 
> We decided to try backporting the fix in #1628 on top of of 1.4.3, but continued to experience consumer lag due to rebalances. Once we reverted to the original 1.4.3 without #1628, the lag stopped. I still see missed heartbeats & rebalances, but they must be happening with less frequency and/or resolving quickly enough to keep consumption steady.
> 
> Is it possible that #1628 contributed to this bug?
> 
> Ping @elukey

Correcting a bit this, my bad for the wrong versioning. We were running 1.4.1 before upgrading to 1.4.6, and we rolled back to 1.4.3. As far as I can see from the changelog, 1.4.3 had several problems with the client consumer code so the locking that we experienced seems something already reported. We still don't explain why 1.4.6 was so unstable. If we manage to reproduce the behavior in a testing environment we'll send more info!

**Comment by jeffwidman** [comment]
@ottomata Can you correct the version numbers in your comment as based on @elukey's comment it appears they're incorrect...??

Also, @elukey I don't see anything in @ottomata's comment about `1.4.6` being unstable, and I don't see anything in your comment other than "it was unstable"... can you explain what symptoms of instability you saw?

`1.4.5` had a lot of bugfixes of prior bugs, resulting a much more stable consumer, but inadvertently introduced a race condition that broke producers under load. 

`1.4.6` fixed that producer issue. I haven't seen other reports of performance/instability problems with `1.4.6`.

**Comment by ottomata** [comment]
I'll just delete my comment since it is a bit irrelevant and incorrect with the wrong version numbers and results.  

Our summary is:

1.4.1 was working fine for us, but another of our users wanted an upgrade of kafka-python to get the fix for #1628.  We upgraded to 1.4.6, and saw the missed heatbeats and rebalances this issue is about.  We incorrectly reverted to 1.4.3 with #1628 backported, and still saw missed heartbeats and rebalances.  So, I was incorrect in guessing that the fix in #1628 may have caused this issue.

Once we reverted all the way back to 1.4.1, the missed heartbeats ceased happening.

>  I haven't seen other reports of performance/instability problems with 1.4.6.
We and others in this thread are reporting that unexpected missed heartbeats happen in 1.4.6.

Thanks, sorry for the confusion!

**Comment by rolisz** [comment]
@jeffwidman I still encounter the problems described in this issue, on 1.4.6. I can provide debug logs if they would useful.

**Comment by jeffwidman** [comment]
> I can provide debug logs if they would useful.

Yes please. In particular, please make sure your log formatter includes timings, as I suspect they will be useless otherwise.

**Comment by elukey** [comment]
@jeffwidman I can provide logs as well, I am testing 1.4.6 in a testing environment with DEBUG log enabled. Usually these are the logs that I see:

```
May 13 23:28:32 deployment-eventlog05 eventlogging-processor@client-side-01[26126]: 2019-05-13 23:28:32,170 [26126] (MainThread) kafka.client [DEBUG] Sending metadata request MetadataRequest_v1(topics=[u'eventlogging-client-side']) to node 2
May 13 23:28:32 deployment-eventlog05 eventlogging-processor@client-side-01[26126]: 2019-05-13 23:28:32,170 [26126] (MainThread) kafka.protocol.parser [DEBUG] Sending request MetadataRequest_v1(topics=[u'eventlogging-client-side'])
May 13 23:28:32 deployment-eventlog05 eventlogging-processor@client-side-01[26126]: 2019-05-13 23:28:32,170 [26126] (MainThread) kafka.conn [DEBUG] <BrokerConnection node_id=2 host=deployment-kafka-jumbo-2.deployment-prep.eqiad.wmflabs:9092 <connected> [IPv4 ('172.16.5.47', 9092)]> Request 117275: MetadataRequest_v1(topics=[u'eventlogging-client-side'])
May 13 23:28:32 deployment-eventlog05 eventlogging-processor@client-side-01[26126]: 2019-05-13 23:28:32,171 [26126] (eventlogging_processor_client_side_00-heartbeat) kafka.coordinator [WARNING] Heartbeat session expired, marking coordinator dead
May 13 23:28:32 deployment-eventlog05 eventlogging-processor@client-side-01[26126]: 2019-05-13 23:28:32,171 [26126] (eventlogging_processor_client_side_00-heartbeat) kafka.coordinator [WARNING] Marking the coordinator dead (node 1) for group eventlogging_processor_client_side_00: Heartbeat session expired.
```

The metadata request before the heartbeat failure seems to happen all the times. How far should I go in the logs to get the culprit of `Heartbeat session expired` ? Is it related to `session_timeout_ms` (we use the default 10s)? Thanks a lot!

**Comment by ottomata** [comment]
@jeffwidman are there other logs we should try and get for ya?

**Comment by tmcqueen-materials** [comment]
We appear to be seeing this issue as well. It seems to happen when the processing code is also multithreaded/multiprocessed and is doing intensive work with large (0.5 MB) messages. Perhaps an issue where the background heartbeat thread gets starved or doesn't know it needs to do something?

**Comment by cvley** [comment]
I have encountered this problem: consumers keep rebalancing when the machine has lots network traffic. I tried to set the `heartbeat_interval_ms` to 1000 ms, rebalancing didn't show in the log, but messages didn't consumed neither.

After I downgraded to `1.3.5`, everything is normal.

Just for reference.

**Comment by vimal3271** [comment]
The below configuration worked for me

'request_timeout_ms': 305000,
'max_poll_interval_ms': 300000,
# Increased to 3 mins from default 10s
'session_timeout_ms': 180000,  # 3mins
# Increased to 1min from default 3s
'heartbeat_interval_ms': 60000,  # 1min
'metadata_max_age_ms': 50000

Plz try out. I am using 1.4.5

**Comment by charettes** [comment]
This is still happening on 1.4.5 and is preventing us from using this library with Python 3.7 because only < 1.4 works appropriately. When using < 1.4 we see frequent heartbeat failures because of rebalancing but more importantly **double consumption of messages** which is highly problematic.

This is also happening when bumping `session.timeout.ms` to larger values which eventually lead to spurious `ValueError: I/O operation on closed epoll fd` errors.

**Comment by piperck** [comment]
We use 1.4.6
It looks like a poll period hang and casue can not send heartbeats.
follow 

> workaround might be to reduce metadata_max_age_ms to something close / equal to your heartbeat_timeout_ms. This will cause more frequent metadata requests, but should unblock the send side when there is no socket data available for reads.

everythings will be fine ....

**Comment by dingyo777** [comment]
Hello guys, I also have same issue with 1.4.6.
Now I've tried to 2nd workaround @dpkp suggested: reduce `metadata_max_age_ms` to `session_timeout_ms`.
Let me report later whether it works or not

**Comment by dingyo777** [comment]
Result!
workaround works well. No heartbeat fails, no rebalance occurs.

**Comment by dpkp** [comment]
I made several improvements to KafkaConsumer and the underlying client that I believe should fix this issue. Please try 1.4.7 and reopen / file issues if this persists!

**Comment by sangmeshcp** [comment]
still hitting this issue with 1.4.7 .. 

[2019-10-02 15:26:21,291] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:21,339] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:21,345] {base.py:746} INFO - Stopping heartbeat thread
[2019-10-02 15:26:21,645] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:24,370] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:24,398] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:24,701] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:27,462] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:27,475] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:27,705] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:30,564] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:30,564] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:30,801] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:33,564] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:33,664] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:33,900] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:36,661] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:36,762] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:37,000] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:39,750] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:39,855] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:40,002] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:42,842] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:42,936] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing
[2019-10-02 15:26:43,094] {base.py:828} WARNING - Heartbeat failed for group caas because it is rebalancing

**Comment by sangmeshcp** [comment]
still hitting this issue with 1.4.7 

[2019-10-02 15:28:24,488] {base.py:828} WARNING - Heartbeat failed for group ca because it is rebalancing
[2019-10-02 15:28:27,587] {base.py:828} WARNING - Heartbeat failed for group ca because it is rebalancing
[2019-10-02 15:28:30,680] {base.py:828} WARNING - Heartbeat failed for group ca because it is rebalancing
[2019-10-02 15:28:33,780] {base.py:828} WARNING - Heartbeat failed for group ca because it is rebalancing
[2019-10-02 15:28:36,882] {base.py:828} WARNING - Heartbeat failed for group ca because it is rebalancing
[2019-10-02 15:28:39,975] {base.py:828} WARNING - Heartbeat failed for group ca because it is rebalancing
[2019-10-02 15:28:43,074] {base.py:828} WARNING - Heartbeat failed for group ca because it is rebalancing
[2019-10-02 15:28:46,165] {base.py:828} WARNING - Heartbeat failed for group ca because it is rebalancing
[2019-10-02 15:28:49,164] {base.py:828} WARNING - Heartbeat failed for group ca because it is rebalancing
[2019-10-02 15:28:52,266] {base.py:828} WARNING - Heartbeat failed for group ca because it is rebalancing

**Comment by jmgpeeters** [comment]
1.4.7 appears to have fixed the issue for us.

**Comment by sangmeshcp** [comment]
> 1.4.7 appears to have fixed the issue for us.

is it will with multiple brokers?

**Comment by elukey** [comment]
> 1.4.7 appears to have fixed the issue for us.

Same thing from my side: I have been testing the new version for 24h now in production and everything went smoothly. Will keep monitor and report back in case anything arise, but so far so good! Thanks!

**Comment by asalem** [comment]
simply you need to reduce the max poll records, default if 500 so you may tweak that to get rid of this warning

**Comment by kennes913** [comment]
I am still having this issue using 1.4.7. Like @sangmeshcp, we have a multiple brokers.

**Comment by naoshadmehta** [comment]
We're seeing this problem with 1.4.6 and 1.4.7. We have 3 broker nodes and 6 consumers per node. Our topic has 18 partitions and for about an hour or so, we see no issues. After an hour or so, we start seeing 

server.log:[2019-10-29 01:28:16,927] INFO [GroupCoordinator 1]: Preparing to rebalance group tasks_group in state PreparingRebalance with old generation 9 (__consumer_offsets-17) (reason: removing member kafka-python-1.4.7-52039bb5-17c5-42d7-8af3-d95f0dbc3f3f on heartbeat expiration) (kafka.coordinator.group.GroupCoordinator)

A few seconds later, we see a new consumer being added:
server.log:[2019-10-29 01:29:00,612] INFO [GroupCoordinator 1]: Preparing to rebalance group tasks_group in state PreparingRebalance with old generation 10 (__consumer_offsets-17) (reason: Adding new member kafka-python-1.4.7-27f5bbac-b1e6-41a9-999b-abb56868e9e2) (kafka.coordinator.group.GroupCoordinator)

After the system gets into this state, we keep seeing remove and add of new members to the consumer group and in a few hours, it reaches a point where the rebalance takes more than 10 minutes to complete. By this time, the lag on the partitions is so high that its impossible to recover from this situation. 

Our broker config:
max.session.timeout.ms: 30000
heartbeat.interval.ms: 10000

Consumer config:
max.session.timeout.ms: 10000
heartbeat.interval.ms: 3000
consumer_timeout_ms: 2000

Everything else is left as default. 

Any suggestions on what we could tweak in the broker or consumer configs to avoid seeing this issue?

**Comment by jmgpeeters** [comment]
Is there any plan to upgrade kafka-python to 1.4.7 on conda-forge? https://anaconda.org/conda-forge/kafka-python/files currently still shows 1.4.6 as the latest.

**Comment by jeffwidman** [comment]
@jmgpeeters please open a new ticket, this one is completely unrelated to condaforge.

**Comment by xiaoma-father** [comment]
i used kafka-python 2.0.1, and I see the warning in log very frequent ,nearly every second, and the client consume message in kafka slow. I solved this question by create many group to consume the topics. firstly I consume all the topics in one group, nearly twelve topics, after finding solution on web, I create five group to consume different function topic ,this problem is solved. I hope I can help you.
